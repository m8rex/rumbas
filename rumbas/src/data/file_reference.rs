use crate::data::input_string::InputString;
use crate::data::optional_overwrite::*;
use crate::data::template::{Value, ValueType};
use serde::Deserialize;
use serde::Serialize;
use std::collections::HashMap;
use std::convert::TryInto;
use std::path::Path;

/// The prefix used to specify a file reference
const FILE_PREFIX: &str = "file:";

macro_rules! file_type {
    (
        $(#[$outer:meta])*
        type $type: ident,
        subtype $subtype: ty
    ) => {
        #[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
        #[serde(from = "String")]
        #[serde(into = "String")]
        $(
            #[$outer]
        )*
        pub struct $type {
            file_name: Option<String>,
            content: Option<$subtype>,
            translated_content: HashMap<String, $subtype>,
            error_message: Option<String>,
        }
        impl RumbasCheck for $type {
            fn check(&self) -> RumbasCheckResult {
                if let Some(e) = &self.error_message {
                    RumbasCheckResult::from_missing(Some(e.clone()))
                } else {
                    RumbasCheckResult::empty()
                }
            }
        }
        impl OptionalOverwrite<$type> for $type {
            fn overwrite(&mut self, _other: &$type) {}
            fn insert_template_value(&mut self, _key: &str, _val: &serde_yaml::Value) {}
        }
        impl_optional_overwrite_value!($type);

        //TODO: error message is not shown if no file found
        impl std::convert::From<String> for $type {
            fn from(s: String) -> Self {
                if s.starts_with(FILE_PREFIX) {
                    if s == FILE_PREFIX {
                        Self {
                            file_name: Some("".to_string()),
                            content: None,
                            translated_content: HashMap::new(),
                            error_message: Some("Missing filename".to_string()),
                        }
                    } else {
                        let relative_file_name = s.split(FILE_PREFIX).collect::<Vec<&str>>()[1];
                        let file_path = Path::new(crate::QUESTIONS_FOLDER).join(relative_file_name);
                        let file_name = file_path.file_name().unwrap().to_str().unwrap(); //TODO
                        if let Some(file_dir) = file_path.parent() {
                            //Look for translation dirs
                            let mut translated_content = HashMap::new();
                            for entry in file_dir.read_dir().expect("read_dir call failed").flatten()
                            // We only care about the ones that are 'Ok'
                            {
                                if let Ok(entry_name) = entry.file_name().into_string() {
                                    //println!("{}", entry_name);
                                    if entry_name.starts_with("locale-") {
                                        let locale = entry_name
                                            .splitn(2, "locale-")
                                            .collect::<Vec<_>>()
                                            .get(1)
                                            .unwrap()
                                            .to_string();
                                        let locale_file_path = file_dir.join(entry_name).join(file_name);
                                        //println!("{}", locale_file_path.display());
                                        if locale_file_path.exists() {
                                            if let Ok(s) = std::fs::read_to_string(&locale_file_path) {
                                                //println!("{}", s);
                                                if let Ok(s) = s.clone().try_into() {
                                                    translated_content.insert(
                                                        locale,
                                                        s,
                                                    );
                                                }
                                                else {
                                                    log::warn!("Failed converting content in {}", locale_file_path.display());
                                                }
                                            }
                                            else {
                                                log::warn!("Failed reading {}", locale_file_path.display());
                                            }
                                        }
                                    }
                                }
                            }

                            let content = std::fs::read_to_string(&file_path)
                                .ok();
                            if content.is_none() && translated_content.is_empty() {
                                Self {
                                    file_name: Some(relative_file_name.to_string()),
                                    content: None,
                                    translated_content: HashMap::new(),
                                    error_message: Some(relative_file_name.to_string()),
                                }
                            } else {
                                if let Some(content) = content {
                                    let content = content.try_into().ok();
                                    if content.is_some() {
                                        Self {
                                            file_name: Some(relative_file_name.to_string()),
                                            content,
                                            translated_content,
                                            error_message: None,
                                        }
                                    }
                                    else {
                                        Self {
                                            file_name: Some(relative_file_name.to_string()),
                                            content: None,
                                            translated_content,
                                            error_message: Some(format!("Failed converting content in {}", file_path.display())),
                                        }
                                    }
                                } else {
                                    Self {
                                        file_name: Some(relative_file_name.to_string()),
                                        content: None,
                                        translated_content,
                                        error_message: None,
                                    }
                                }
                            }
                        } else {
                            Self {
                                file_name: Some(relative_file_name.to_string()),
                                content: None,
                                translated_content: HashMap::new(),
                                error_message: Some(relative_file_name.to_string()),
                            }
                        }
                    }
                } else {
                    Self::s(&s)
                }
            }
        }

        // Currently only implemented for conversion from numbas to rumbas: so not with translations or
        // file references
        impl std::convert::From<$type> for String {
            fn from(fs: $type) -> Self {
                if fs.file_name.is_some() || !fs.translated_content.is_empty() || fs.content.is_none() {
                    panic!("Deserializing FileRef only supported when plain String")
                }
                fs.content.unwrap().into()
            }
        }

        impl $type {
            pub fn get_content(&self, locale: &str) -> String {
                if let Some(c) = self.translated_content.get(locale) {
                    return c.clone().into();
                }
                if let Some(c) = &self.content {
                    return c.clone().into();
                }
                panic!("Missing translation for locale {}", locale); //TODO
            }
            pub fn s(content: &str) -> $type {
                let content = content.to_string().try_into();
                let error_message = if let Err(e) = content {
                    Some(format!("Invalid file content: {}", e))
                } else { None };
                $type {
                    file_name: None,
                    content: content.ok(),
                    translated_content: HashMap::new(),
                    error_message,
                }
            }
        }
    }
}

file_type! {
    /// A string that has to be read from a file.
    ///
    /// Specified by a string starting with [FILE_PREFIX].
    type FileString,
    subtype InputString
}

file_type! {
    /// A JME string that has to be read from a file.
    ///
    /// Specified by a string starting with [FILE_PREFIX].
    type JMEFileString,
    subtype String
}

file_type! {
    /// An embraced JME string that has to be read from a file.
    ///
    /// Specified by a string starting with [FILE_PREFIX].
    type EmbracedJMEFileString,
    subtype String
}
