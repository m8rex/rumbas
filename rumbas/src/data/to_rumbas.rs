use crate::data::question_part::question_part::QuestionPart;
use crate::data::translatable::ContentAreaTranslatableString;
use crate::data::translatable::EmbracedJMETranslatableString;
use crate::data::translatable::JMETranslatableString;
use crate::data::translatable::TranslatableString;
use numbas::defaults::DEFAULTS;
use numbas::jme::{ContentAreaString, EmbracedJMEString, JMEString};
use std::convert::TryInto;

pub trait ToRumbas<RumbasType>: Clone {
    fn to_rumbas(&self) -> RumbasType;
}

macro_rules! impl_to_rumbas {
    ($($type: ty$([$($gen: tt), *])?), *) => {
        $(
        impl$(< $($gen : Clone),* >)? ToRumbas<$type> for $type {
            fn to_rumbas(&self) -> $type {
                self.clone()
            }
        }
        )*
    };
}

impl_to_rumbas!(bool, f64, usize, [f64; 2]);
impl_to_rumbas!(numbas::exam::Primitive);
impl_to_rumbas!(numbas::jme::JMEString);
impl_to_rumbas!(numbas::jme::EmbracedJMEString);
impl_to_rumbas!(numbas::jme::ContentAreaString);

impl<T, O: ToRumbas<T>> ToRumbas<Vec<T>> for Vec<O> {
    fn to_rumbas(&self) -> Vec<T> {
        self.iter().map(|item| item.to_rumbas()).collect()
    }
}

impl<T, O: ToRumbas<T>> ToRumbas<Option<T>> for Option<O> {
    fn to_rumbas(&self) -> Option<T> {
        self.clone().map(|item| item.to_rumbas())
    }
}

impl ToRumbas<TranslatableString> for String {
    fn to_rumbas(&self) -> TranslatableString {
        self.clone().into()
    }
}

impl ToRumbas<JMETranslatableString> for JMEString {
    fn to_rumbas(&self) -> JMETranslatableString {
        self.clone().into()
    }
}

impl ToRumbas<EmbracedJMETranslatableString> for EmbracedJMEString {
    fn to_rumbas(&self) -> EmbracedJMETranslatableString {
        self.clone().into()
    }
}

impl ToRumbas<ContentAreaTranslatableString> for ContentAreaString {
    fn to_rumbas(&self) -> ContentAreaTranslatableString {
        self.clone().into()
    }
}

pub fn extract_part_common_marks(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> numbas::exam::Primitive {
    pd.marks
        .clone()
        .unwrap_or(numbas::exam::Primitive::Natural(DEFAULTS.part_common_marks))
}

pub fn extract_part_common_prompt(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> ContentAreaTranslatableString {
    pd.prompt
        .clone()
        .unwrap_or_else(|| "".to_string().try_into().unwrap())
        .to_rumbas()
}

pub fn extract_part_common_use_custom_name(pd: &numbas::exam::ExamQuestionPartSharedData) -> bool {
    pd.use_custom_name
        .unwrap_or(DEFAULTS.part_common_use_custom_name)
}

pub fn extract_part_common_custom_name(pd: &numbas::exam::ExamQuestionPartSharedData) -> String {
    pd.custom_name.clone().unwrap_or_default()
}

pub fn extract_part_common_steps_penalty(pd: &numbas::exam::ExamQuestionPartSharedData) -> usize {
    pd.steps_penalty
        .unwrap_or(DEFAULTS.part_common_steps_penalty)
}

pub fn extract_part_common_enable_minimum_marks(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> bool {
    pd.enable_minimum_marks
        .unwrap_or(DEFAULTS.part_common_enable_minimum_marks)
}

pub fn extract_part_common_minimum_marks(pd: &numbas::exam::ExamQuestionPartSharedData) -> usize {
    pd.minimum_marks
        .unwrap_or(DEFAULTS.part_common_minimum_marks)
}

pub fn extract_part_common_show_correct_answer(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> bool {
    pd.show_correct_answer
}

pub fn extract_part_common_show_feedback_icon(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> bool {
    pd.show_feedback_icon
        .unwrap_or(DEFAULTS.part_common_show_feedback_icon)
}

pub fn extract_part_common_adaptive_marking_penalty(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> usize {
    pd.adaptive_marking_penalty
        .unwrap_or(DEFAULTS.part_common_adaptive_marking_penalty)
}

pub fn extract_part_common_extend_base_marking_algorithm(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> bool {
    pd.extend_base_marking_algorithm
        .unwrap_or(DEFAULTS.part_common_extend_base_marking_algorithm)
}

pub fn extract_part_common_steps(
    pd: &numbas::exam::ExamQuestionPartSharedData,
) -> Vec<QuestionPart> {
    pd.steps
        .clone()
        .unwrap_or_default()
        .into_iter()
        .map(|s| s.to_rumbas())
        .collect()
}
