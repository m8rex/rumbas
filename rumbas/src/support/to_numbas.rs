use crate::support::optional_overwrite::*;

pub type NumbasResult<T> = Result<T, RumbasCheckResult>;

pub trait ToNumbas<NumbasType>: Clone + RumbasCheck {
    /// Method that safely converts a type to another (probably numbas) type
    fn to_numbas_safe(&self, locale: &str) -> NumbasResult<NumbasType> {
        let check = self.check(locale);
        if check.is_empty() {
            Ok(self.to_numbas(locale))
        } else {
            Err(check)
        }
    }
    /// Method that converts a type to another type
    /// This method assumes that it is called by a function that is initially called from `to_numbas_safe`
    fn to_numbas(&self, locale: &str) -> NumbasType;
    fn to_numbas_with_name(&self, locale: &str, _name: String) -> NumbasType {
        self.to_numbas(locale)
    }
}

impl_to_numbas!(String, bool, f64, usize);
impl_to_numbas!(numbas::jme::JMEString);
impl_to_numbas!(numbas::jme::EmbracedJMEString);
impl_to_numbas!(numbas::jme::ContentAreaString);

impl<S, O: ToNumbas<S>> ToNumbas<Vec<S>> for Vec<O> {
    fn to_numbas(&self, locale: &str) -> Vec<S> {
        let mut v = Vec::new();
        for item in self.iter() {
            v.push(item.to_numbas(locale));
        }
        v
    }
}

impl<K: Clone + std::hash::Hash + std::cmp::Eq, S, O: ToNumbas<S>>
    ToNumbas<std::collections::HashMap<K, S>> for std::collections::HashMap<K, O>
where
    std::collections::HashMap<K, O>: RumbasCheck,
{
    fn to_numbas(&self, locale: &str) -> std::collections::HashMap<K, S> {
        self.iter()
            .map(|(k, v)| (k.to_owned(), v.to_numbas(locale)))
            .collect()
    }
}

impl<AA, A: ToNumbas<AA>, BB, B: ToNumbas<BB>> ToNumbas<(AA, BB)> for (A, B)
where
    (A, B): RumbasCheck,
{
    fn to_numbas(&self, locale: &str) -> (AA, BB) {
        (self.0.to_numbas(locale), self.1.to_numbas(locale))
    }
}

impl<AA, A: ToNumbas<AA>> ToNumbas<[AA; 2]> for [A; 2]
where
    [A; 2]: RumbasCheck,
{
    fn to_numbas(&self, locale: &str) -> [AA; 2] {
        [self[0].to_numbas(locale), self[1].to_numbas(locale)]
    }
}

impl ToNumbas<numbas::exam::SafeFloat> for f64 {
    fn to_numbas(&self, _locale: &str) -> numbas::exam::SafeFloat {
        (*self).into()
    }
}

impl ToNumbas<numbas::exam::SafeNatural> for usize {
    fn to_numbas(&self, _locale: &str) -> numbas::exam::SafeNatural {
        (*self).into()
    }
}

macro_rules! impl_to_numbas {
    ($($type: ty), *) => {
        $(
        impl ToNumbas<$type> for $type {
            fn to_numbas(&self, _locale: &str) -> $type {
                self.clone()
            }
        }
        )*
    };
}

pub(crate) use impl_to_numbas;
