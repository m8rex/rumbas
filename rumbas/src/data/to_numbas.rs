use crate::data::optional_overwrite::{Noneable, OptionalOverwrite};
use crate::data::template::{Value, ValueType};

pub type NumbasResult<T> = Result<T, Vec<String>>;

pub trait ToNumbas: Clone {
    type NumbasType;
    fn to_numbas(&self, locale: &String) -> NumbasResult<Self::NumbasType>;
    fn to_numbas_with_name(
        &self,
        locale: &String,
        _name: String,
    ) -> NumbasResult<Self::NumbasType> {
        self.to_numbas(&locale)
    }
}

impl<T: ToNumbas + OptionalOverwrite> ToNumbas for Value<T> {
    type NumbasType = <T as ToNumbas>::NumbasType;
    fn to_numbas(&self, locale: &String) -> NumbasResult<Self::NumbasType> {
        match &self.0 {
            Some(ValueType::Normal(val)) => {
                let empty_fields = val.empty_fields();
                if empty_fields.is_empty() {
                    Ok(val.to_numbas(&locale).unwrap())
                } else {
                    Err(empty_fields)
                }
            }
            Some(ValueType::Template(ts)) => Err(vec![ts.yaml()]),
            None => Err(vec!["".to_string()]),
        }
    }
}

impl<T: ToNumbas + OptionalOverwrite> ToNumbas for Noneable<T> {
    type NumbasType = Option<<T as ToNumbas>::NumbasType>;
    fn to_numbas(&self, locale: &String) -> NumbasResult<Self::NumbasType> {
        match self {
            Noneable::NotNone(val) => {
                let empty_fields = val.empty_fields();
                if empty_fields.is_empty() {
                    Ok(Some(val.clone().to_numbas(&locale).unwrap()))
                } else {
                    Err(empty_fields)
                }
            }
            _ => Ok(None),
        }
    }
}

macro_rules! impl_to_numbas {
    ($($type: ty), *) => {
        $(
        impl ToNumbas for $type {
            type NumbasType = $type;
            fn to_numbas(&self, _locale: &String) -> NumbasResult<Self::NumbasType> {
                Ok(self.clone())
            }
        }
        )*
    };
}

impl_to_numbas!(String, bool, f64, usize, [f64; 2]);

impl<O: ToNumbas> ToNumbas for Vec<O> {
    type NumbasType = Vec<O::NumbasType>;
    fn to_numbas(&self, locale: &String) -> NumbasResult<Self::NumbasType> {
        let mut v = Vec::new();
        for item in self.into_iter() {
            v.push(item.to_numbas(&locale)?);
        }
        Ok(v)
    }
}
