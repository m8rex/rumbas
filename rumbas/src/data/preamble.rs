use crate::data::file_reference::FileString;
use crate::data::optional_overwrite::{Noneable, OptionalOverwrite};
use crate::data::to_numbas::{NumbasResult, ToNumbas};
use serde::{Deserialize, Serialize};

// TODO: maybe translatable? So different text's can be set for different languages?
optional_overwrite! {
    Preamble,
    js: FileString,
    css: FileString
}

impl ToNumbas for Preamble {
    type NumbasType = numbas::exam::Preamble;
    fn to_numbas(&self, _locale: &String) -> NumbasResult<numbas::exam::Preamble> {
        let empty_fields = self.empty_fields();
        if empty_fields.is_empty() {
            Ok(numbas::exam::Preamble::new(
                self.js.clone().unwrap().get_content(),
                self.css.clone().unwrap().get_content(),
            ))
        } else {
            Err(empty_fields)
        }
    }
}
