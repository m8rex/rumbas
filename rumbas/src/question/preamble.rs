use crate::support::file_reference::FileString;
use crate::support::optional_overwrite::*;
use crate::support::template::{Value, ValueType};
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

// TODO: maybe translatable? So different text's can be set for different languages?
optional_overwrite! {
    pub struct Preamble {
        /// The JavaScript to add to the outputfiles
        js: FileString,
        /// The CSS to add to the outputfiles
        css: FileString
    }
}

impl ToNumbas<numbas::exam::Preamble> for Preamble {
    fn to_numbas(&self, locale: &str) -> numbas::exam::Preamble {
        numbas::exam::Preamble {
            js: self.js.to_numbas(locale),
            css: self.css.to_numbas(locale),
        }
    }
}

impl ToRumbas<Preamble> for numbas::exam::Preamble {
    fn to_rumbas(&self) -> Preamble {
        Preamble {
            js: self.js.to_rumbas(),
            css: self.css.to_rumbas(),
        }
    }
}
