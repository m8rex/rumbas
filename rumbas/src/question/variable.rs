use crate::support::file_reference::FileString;
use crate::support::to_numbas::ToNumbas;
use crate::support::to_rumbas::ToRumbas;
use regex::Regex;
use rumbas_support::preamble::*;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

pub const UNGROUPED_GROUP: &str = "Ungrouped variables";

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "VariableRepresentationInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
#[serde(untagged)]
pub enum VariableRepresentation {
    ListOfStrings(Vec<String>),
    ListOfNumbers(Vec<f64>),
    Long(Box<Variable>),
    Number(f64),
    Other(VariableStringRepresentation),
}

impl ToNumbas<numbas::question::variable::Variable> for VariableRepresentation {
    fn to_numbas_with_name(
        &self,
        locale: &str,
        name: String,
    ) -> numbas::question::variable::Variable {
        self.to_variable().to_numbas_with_name(locale, name)
    }
    fn to_numbas(&self, _locale: &str) -> numbas::question::variable::Variable {
        panic!(
            "{}",
            "Should not happen, don't call this method Missing name".to_string(),
        )
    }
}

impl ToRumbas<VariableRepresentation> for numbas::question::variable::Variable {
    fn to_rumbas(&self) -> VariableRepresentation {
        VariableRepresentation::Long(Box::new(Variable {
            definition: self.definition.to_rumbas(),
            description: self.description.to_rumbas(),
            template_type: self.template_type.to_rumbas(),
            group: self.group.to_rumbas(),
        }))
    }
}

impl VariableRepresentation {
    pub fn to_variable(&self) -> Variable {
        match self {
            VariableRepresentation::ListOfStrings(l) => Variable::ungrouped(
                VariableTemplateType::ListOfStrings,
                &serde_json::to_string(&l).unwrap(),
            ),
            VariableRepresentation::ListOfNumbers(l) => Variable::ungrouped(
                VariableTemplateType::ListOfNumbers,
                &serde_json::to_string(&l).unwrap(),
            ),
            VariableRepresentation::Long(v) => *(v.clone()),
            VariableRepresentation::Number(n) => {
                Variable::ungrouped(VariableTemplateType::Number, &n.to_string())
            }
            VariableRepresentation::Other(o) => match o {
                VariableStringRepresentation::Anything(s) => {
                    Variable::ungrouped(VariableTemplateType::Anything, &s)
                }
                VariableStringRepresentation::Range(r) => {
                    Variable::ungrouped(VariableTemplateType::Range, &r.as_range())
                }
                VariableStringRepresentation::RandomRange(r) => {
                    Variable::ungrouped(VariableTemplateType::RandomRange, &r.as_random_range())
                }
            },
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "VariableStringRepresentationInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
#[serde(from = "String")]
pub enum VariableStringRepresentation {
    Anything(String),
    Range(RangeData),
    RandomRange(RangeData),
}
/* TODO remove?
impl JsonSchema for VariableStringRepresentation {
    fn schema_name() -> String {
        "VariableStringRepresentation".to_owned()
    }

    fn json_schema(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
        schemars::schema::SchemaObject {
            instance_type: Some(schemars::schema::InstanceType::String.into()),
            ..Default::default()
        }
        .into()
    }
}*/

impl std::convert::From<String> for VariableStringRepresentation {
    fn from(s: String) -> Self {
        if let Some(r) = RangeData::try_from_range(&s) {
            VariableStringRepresentation::Range(r)
        } else if let Some(r) = RangeData::try_from_random_range(&s) {
            VariableStringRepresentation::RandomRange(r)
        } else {
            VariableStringRepresentation::Anything(s)
        }
    }
}

impl std::convert::From<String> for VariableStringRepresentationInput {
    fn from(s: String) -> Self {
        let v = VariableStringRepresentation::from(s);
        Self::from_normal(v)
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "RangeDataInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema, PartialEq)]
pub struct RangeData {
    pub from: f64,
    pub to: f64,
    pub step: f64,
}

impl RangeData {
    pub fn try_from_range(s: &str) -> Option<RangeData> {
        let re = Regex::new(r"^(\d+(?:\.\d*)?)\s*\.\.\s*(\d+(?:\.\d*)?)(\#(\d+(?:\.\d*)?))?$")
            .expect("It to be a valid regex");
        if let Some(c) = re.captures(s) {
            return Some(RangeData {
                from: c.get(1).unwrap().as_str().parse().unwrap(),
                to: c.get(2).unwrap().as_str().parse().unwrap(),
                step: c.get(4).map(|m| m.as_str()).unwrap_or("1").parse().unwrap(),
            });
        }
        None
    }
    pub fn as_range(&self) -> String {
        format!("{} .. {}#{}", self.from, self.to, self.step)
    }
    pub fn try_from_random_range(s: &str) -> Option<RangeData> {
        if s.starts_with("random(") && s.ends_with(')') {
            return RangeData::try_from_range(&s[7..s.len() - 1].to_string());
        }
        None
    }
    pub fn as_random_range(&self) -> String {
        format!("random({})", self.as_range())
    }
}

#[cfg(test)]
mod test {
    use super::*;

    macro_rules! i {
        ($expr: expr) => {
            RangeDataInput::from_normal($expr)
        };
    }

    #[test]
    fn range_ints() {
        let s = "2 .. 10#1".to_string();
        assert_eq!(
            Some(i!(RangeData {
                from: 2.0,
                to: 10.0,
                step: 1.0
            })),
            RangeData::try_from_range(&s).map(|a| i!(a))
        );
        assert_eq!(None, RangeData::try_from_random_range(&s).map(|a| i!(a)));
    }

    #[test]
    fn random_range_ints() {
        let s = "random(2 .. 10#2)".to_string();
        assert_eq!(
            Some(i!(RangeData {
                from: 2.0,
                to: 10.0,
                step: 2.0
            })),
            RangeData::try_from_random_range(&s).map(|a| i!(a))
        );
        assert_eq!(None, RangeData::try_from_range(&s).map(|a| i!(a)));
    }

    #[test]
    fn random_range_ints_without_step() {
        let s = "random(2..10)".to_string();
        assert_eq!(
            Some(i!(RangeData {
                from: 2.0,
                to: 10.0,
                step: 1.0
            })),
            RangeData::try_from_random_range(&s).map(|a| i!(a))
        );
        assert_eq!(None, RangeData::try_from_range(&s).map(|a| i!(a)));
    }

    #[test]
    fn range_floats() {
        let s = "2.1 .. 10.5#1.99".to_string();
        assert_eq!(
            Some(i!(RangeData {
                from: 2.1,
                to: 10.5,
                step: 1.99
            })),
            RangeData::try_from_range(&s).map(|a| i!(a))
        );
        assert_eq!(None, RangeData::try_from_random_range(&s).map(|a| i!(a)));
    }

    #[test]
    fn random_range_floats() {
        let s = "random(20.222 .. 100.0#19.57)".to_string();
        assert_eq!(
            Some(i!(RangeData {
                from: 20.222,
                to: 100.0,
                step: 19.57
            })),
            RangeData::try_from_random_range(&s).map(|a| i!(a))
        );
        assert_eq!(None, RangeData::try_from_range(&s).map(|a| i!(a)));
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "VariableInput")]
#[derive(Serialize, Deserialize, Debug, Clone, JsonSchema)]
pub struct Variable {
    pub definition: FileString, //TODO: definition dependant of template type, for random_range: start, end and step instead
    pub description: String,
    pub template_type: VariableTemplateType,
    pub group: String, //TODO "Ungrouped variables" -> real optional? if not -> ungrouped?
}

impl ToNumbas<numbas::question::variable::Variable> for Variable {
    fn to_numbas_with_name(
        &self,
        locale: &str,
        name: String,
    ) -> numbas::question::variable::Variable {
        numbas::question::variable::Variable {
            name,
            definition: self.definition.to_numbas(locale),
            description: self.description.to_numbas(locale),
            template_type: self.template_type.to_numbas(locale),
            group: self.group.to_numbas(locale),
            can_override: false, // Don't support overriding variables (yet?)
        }
    }
    fn to_numbas(&self, _locale: &str) -> numbas::question::variable::Variable {
        panic!(
            "{}",
            "Should not happen, don't call this method Missing name".to_string(),
        )
    }
}

impl Variable {
    fn ungrouped(template_type: VariableTemplateType, definition: &str) -> Variable {
        Variable {
            template_type,
            definition: definition.to_string().to_rumbas(),
            description: String::new().to_rumbas(),
            group: UNGROUPED_GROUP.to_string().to_rumbas(),
        }
    }
}

#[derive(Input, Overwrite, RumbasCheck)]
#[input(name = "VariableTemplateTypeInput")]
/// The different template_types for a variable
#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum VariableTemplateType {
    /// Not specified
    Anything, // TODO: allow this?
    /// A list of numbers
    ListOfNumbers,
    /// A list of strings
    ListOfStrings,
    /// A long string
    LongString,
    /// A number
    Number,
    /// A random number from a range
    RandomRange,
    /// A range
    Range,
    /// A string
    r#String,
}

impl ToNumbas<numbas::question::variable::VariableTemplateType> for VariableTemplateType {
    fn to_numbas(&self, _locale: &str) -> numbas::question::variable::VariableTemplateType {
        match self {
            VariableTemplateType::Anything => {
                numbas::question::variable::VariableTemplateType::Anything
            }
            VariableTemplateType::ListOfNumbers => {
                numbas::question::variable::VariableTemplateType::ListOfNumbers
            }
            VariableTemplateType::ListOfStrings => {
                numbas::question::variable::VariableTemplateType::ListOfStrings
            }
            VariableTemplateType::LongString => {
                numbas::question::variable::VariableTemplateType::LongString
            }
            VariableTemplateType::Number => {
                numbas::question::variable::VariableTemplateType::Number
            }
            VariableTemplateType::Range => numbas::question::variable::VariableTemplateType::Range,
            VariableTemplateType::RandomRange => {
                numbas::question::variable::VariableTemplateType::RandomRange
            }
            VariableTemplateType::r#String => {
                numbas::question::variable::VariableTemplateType::r#String
            }
        }
    }
}

impl ToRumbas<VariableTemplateType> for numbas::question::variable::VariableTemplateType {
    fn to_rumbas(&self) -> VariableTemplateType {
        match self {
            numbas::question::variable::VariableTemplateType::Anything => {
                VariableTemplateType::Anything
            }
            numbas::question::variable::VariableTemplateType::ListOfNumbers => {
                VariableTemplateType::ListOfNumbers
            }
            numbas::question::variable::VariableTemplateType::ListOfStrings => {
                VariableTemplateType::ListOfStrings
            }
            numbas::question::variable::VariableTemplateType::LongString => {
                VariableTemplateType::LongString
            }
            numbas::question::variable::VariableTemplateType::Number => {
                VariableTemplateType::Number
            }
            numbas::question::variable::VariableTemplateType::RandomRange => {
                VariableTemplateType::RandomRange
            }
            numbas::question::variable::VariableTemplateType::Range => VariableTemplateType::Range,
            numbas::question::variable::VariableTemplateType::r#String => {
                VariableTemplateType::r#String
            }
        }
    }
}
