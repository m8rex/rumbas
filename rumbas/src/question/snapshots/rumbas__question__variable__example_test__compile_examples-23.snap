---
source: rumbas/src/question/variable.rs
expression: "&example"

---
definition: "template:template_key"
description: other text
template_type: list_of_numbers
group: other text

